staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queueset
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apiserver/pkg/util/flowcontrol/counter"
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/apiserver/pkg/util/shufflesharding"
+	"k8s.io/klog"
+)
+
+// queueSetFactoryImpl implements the QueueSetFactory interface
+// queueSetFactoryImpl makes QueueSet objects.
+type queueSetFactoryImpl struct {
+	counter counter.GoRoutineCounter
+
+	clk clock.PassiveClock
+}
+
+// NewQueueSetFactory creates a new QueueSetFactory object
+func NewQueueSetFactory(clk clock.PassiveClock, counter counter.GoRoutineCounter) fq.QueueSetFactory {
+	return &queueSetFactoryImpl{
+		counter: counter,
+		clk:     clk,
+	}
+}
+
+// NewQueueSet creates a new QueueSet object
+// There is a new QueueSet created for each priority level.
+func (qsf queueSetFactoryImpl) NewQueueSet(name string, concurrencyLimit,
+	desiredNumQueues, queueLengthLimit, handSize int,
+	requestWaitLimit time.Duration) (fq.QueueSet, error) {
+	return newQueueSetImpl(name, concurrencyLimit, desiredNumQueues,
+		queueLengthLimit, handSize, requestWaitLimit, qsf.clk, qsf.counter)
+}
+
+// queueSetImpl is a fair queuing implementation designed with three major differences:
+// 1) dispatches requests to be served rather than packets to be transmitted
+// 2) serves multiple requests at once
+// 3) a request's service time is not known until it finishes
+// implementation of:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+type queueSetImpl struct {
+	lock                 sync.Mutex
+	name                 string
+	counter              counter.GoRoutineCounter
+	queues               []*fq.Queue
+	clk                  clock.PassiveClock
+	vt                   float64
+	estimatedServiceTime float64
+	lastRealTime         time.Time
+	robinIdx             int
+	// numRequestsEnqueued is the number of requests currently enqueued
+	// (eg: incremeneted on Enqueue, decremented on Dequue)
+	numRequestsEnqueued int
+	concurrencyLimit    int
+	desiredNumQueues    int
+	queueLengthLimit    int
+	requestWaitLimit    time.Duration
+	emptyHandler        fq.EmptyHandler
+	dealer              *shufflesharding.Dealer
+}
+
+// initQueues is a helper method for initializing an array of n queues
+func initQueues(n, baseIndex int) []*fq.Queue {
+	fqqueues := make([]*fq.Queue, n)
+	for i := 0; i < n; i++ {
+		fqqueues[i] = &fq.Queue{Index: baseIndex + i, Requests: make([]*fq.Request, 0)}
+
+	}
+	return fqqueues
+}
+
+// newQueueSetImpl creates a new queueSetImpl from passed in parameters
+func newQueueSetImpl(name string, concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int,
```

Comment:
I personally would collect the options into a separate named struct; reduces chance for transposition errors, cuts down on the boilerplate when constructing the queueSetImpl